### DISCUSSION TITLE
Inquiry Regarding Custom Vehicle Controller Implementation

### DISCUSSION THREAD

**Discussion by cooldelay** [openTCS/opentcs#172] [General]
Dear openTCS team,

I am currently working on a project that involves extending and customizing the VehicleControllerPool functionality in OpenTCS. As part of this effort, I have created a custom implementation of DefaultVehicleControllerPoolV2 to better suit my project requirements.

However, I have encountered some challenges with the VehicleControllerFactory interface, which mandates the creation of a DefaultVehicleController instance through the createVehicleController method. I would like to implement our own custom vehicle controller, but the current factory method restricts this by only allowing the creation of DefaultVehicleController, within the file located at [open-tcs\openTCS-Kernel\src\main\java\org\opentcs\kernel\vehicles\VehicleControllerFactory.java](https://github.com/openTCS/opentcs/blob/master/opentcs-kernel/src/main/java/org/opentcs/kernel/vehicles/VehicleControllerFactory.java#L20), between lines 11 and 21.

Could you kindly provide guidance on the following points:

Is it possible to modify or extend the VehicleControllerFactory interface to support custom implementations of VehicleController?
Are there any best practices or official recommendations for extending the DefaultVehicleController functionality while maintaining compatibility with the current OpenTCS framework?
Could you advise on the best approach to implement a flexible factory that allows us to create our own controller, without diverging from OpenTCS's architecture?
I would greatly appreciate your feedback on these questions. Any guidance or resources you could share would be invaluable to me as I work on extending the controller functionality.

Thank you for your time and support.

**Comment by swltr** [comment]
@cooldelay As mentioned in https://github.com/orgs/openTCS/discussions/169#discussioncomment-11029569, `VehicleControllerFactory` is not part of the openTCS API. It is part of the kernel implementation. Extending or modifying this interface or extending `DefaultVehicleController` is not supported. What is part of the API is the interface `VehicleControllerPool`, so one possible way to achieve what you're trying to do is to bind your own implementation of `VehicleControllerPool` and work with that.
